### PR TITLE
[Gen] Optimize marshalling on some struct pinvokes

### PR DESCRIPTION
**Pull request by Therzok**
This changes native signatures to use by-ref passing instead of having to marshal structs.
The idea behind this is to avoid one extra allocation + free, box operation from StructureToPtr and reconstruction of the value after an unmanaged call
MarshalReturnType from VirtualMethod was using the wrong delegation target.
Also, now that the pango_glyph_item_free generated method has a different signature, bind a manual one so we can build and use it.

for 874404d2f0f41eec4cc0cf2e580194fcfd7f5492: https://gist.github.com/Therzok/18323659503725ad5bb5501477e0495f

### CODE REVIEW

**Comment by Therzok** [comment]
cc @alanmcgovern

**Comment by Therzok** [comment]
for b007e646948ba879ad6467cd216ca8b9f8977482: https://gist.github.com/Therzok/4e76d191033cc5f8fd5dde9b79ea0a1c

**Comment by Therzok** [comment]
between 2c8807e and 449d4e2 (this actually takes into account all the changes, as the previous variant might've been corrupting memory) https://gist.github.com/Therzok/f3582e15eb75379077af0bba6edcc3dc

**Comment by Therzok** [comment]
308d4d3d3d99ee8405c84ea75d093b551472f42a: https://gist.github.com/Therzok/955df4fda19dbbe3f260d3a9fbc895fd

**Comment by Therzok** [comment]
2bd3e23: https://gist.github.com/Therzok/48e4290a942ec7067a4938627877c9d5

**Comment by Therzok** [comment]
For those not faint of heart:
https://gist.github.com/Therzok/9dbe15bf89d8e8ab8f7018474c34faad

 945 files changed, 1307 insertions(+), 3004 deletions(-)

**Comment by alanmcgovern** [comment]
In general LGTM. However we should land it after C9 is fully branched. We're getting a bit close to the branch dates for me to be super comfortable landing big changes like this :)

File: glib/IntPtrEqualityComparer.cs [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 namespace GLib
 {
-	class IntPtrEqualityComparer : IEqualityComparer<IntPtr>
+	public class IntPtrEqualityComparer : IEqualityComparer<IntPtr>
```
Comment (Therzok) [inline glib/IntPtrEqualityComparer.cs]: I'll dup the impl in gtk

**Comment by Therzok** [comment]
For e150f445bdb7b1d6faa846088ceb7377e28e7ed4: https://gist.github.com/Therzok/4017a9a6725f1be075e7911660aa6fa5

File: glib/SignalClosure.cs [hunk 2]
```diff
@@ -150,26 +152,18 @@ static void MarshalCallback (IntPtr raw_closure, IntPtr return_val, uint n_param
 
 				SignalArgs args = Activator.CreateInstance (closure.args_type, new object [0]) as SignalArgs;
 				args.Args = new object [n_param_vals - 1];
-				GLib.Value[] vals = new GLib.Value [n_param_vals - 1];
-				int valueSize = Marshal.SizeOf (typeof (Value));
 				for (int i = 1; i < n_param_vals; i++) {
-					IntPtr ptr = new IntPtr (param_values.ToInt64 () + i * valueSize);
-					vals [i - 1] = (Value) Marshal.PtrToStructure (ptr, typeof (Value));
-					args.Args [i - 1] = vals [i - 1].Val;
+					args.Args [i - 1] = param_values [i].Val;
```
Comment (Therzok) [inline glib/SignalClosure.cs]: Oops

File: glib/FastActivator.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		static FastCreateObject FastCtor (Type t)
+		{
+			FastCreateObject method;
+			if (!cache.TryGetValue (t, out method)) {
+				var param = Expression.Parameter (typeof (IntPtr));
+				var newExpr = Expression.New (
+					t.GetConstructor (flags, null, new [] { typeof (IntPtr) }, new ParameterModifier [0]),
+					param
+				);
+				cache [t] = method = (FastCreateObject)Expression.Lambda (typeof (FastCreateObject), newExpr, param).Compile ();
```
Comment (alanmcgovern) [inline glib/FastActivator.cs]: This seems like it'd be thread unsafe. Shouldn't we be locking the cache to protect it from concurrent writers?

Comment (Therzok) [inline glib/FastActivator.cs]: No, we shouldn't, as the objects should only be created on the main thread.

Comment (alanmcgovern) [inline glib/FastActivator.cs]: `GLib itself is internally completely thread-safe (all global data is automatically locked),`

File: glib/Object.cs [hunk 4]
```diff
@@ -41,6 +41,8 @@ public class Object : IWrapper, IDisposable {
 		static object lockObject = new object ();
 		static List<ToggleRef> PendingDestroys = new List<ToggleRef> ();
 		static bool idle_queued;
+		protected static IntPtr[] ZeroPtrArray = new IntPtr [0];
+		protected static Value[] ZeroValueArray = new Value [0];
```
Comment (alanmcgovern) [inline glib/Object.cs]: This is a public API change as well. Would `internal` be a better option here?

Comment (Therzok) [inline glib/Object.cs]: It's an addition, not a removal?

Comment (Therzok) [inline glib/Object.cs]: I forgot about that :/

File: glib/FastActivator.cs (hunk 3 above)
Comment (Therzok) [inline glib/FastActivator.cs]: Fixed thread safety.

File: glib/Object.cs (hunk 4 above)
Comment (Therzok) [inline glib/Object.cs]: Fixed this too.

**Comment by alanmcgovern** [comment]
Lgtm

**Pull request merged**